report: close drop down menu when focus is lost

johnemau opened the pull request:
**Summary**
#### Problem
The report drop down menu does not close when the escape key
is pressed in the devtools.

***Gif of bug: menu does not close when escape key is pressed and console opens***
![devtools-menu-console-esc-bug](https://user-images.githubusercontent.com/309310/72118994-cd472180-3307-11ea-91e0-767d543d8c70.gif)

This creates a challenge for customer relying on the keyboard that wish
to dismiss the drop down.

This also does not align with the experience of the WAI ARIA practice
navigation menu button example [1].

#### Root cause
The escape key event is captured and not propagated by the
devtools to toggle the console drawer.

#### Solution
Close the menu when focus moves out of the menu.

This better aligns the menu with the Navigation menu button example for
which its interaction design is based on [1].

**Code Change**
* Add a menu focus out handler to close the drop down menu
* Unfortunately JSDOM does not support focusout events so JSDOM integration tests
  are not possible [2], instead I have added unit test for the onMenuFocusOut handler.
* A TypeScript upgrade is needed to better support the focusout event [3].

***Gif of bug fixed: menu does close when escape key is pressed and console opens***
![devtools-menu-console-esc-bug-fixed](https://user-images.githubusercontent.com/309310/72119010-d89a4d00-3307-11ea-9c8a-e758c06790e7.gif)

***Gif of menu escape key closing the menu***
![devtools-menu-console-esc-bug-fixed-demo-esc](https://user-images.githubusercontent.com/309310/72119043-efd93a80-3307-11ea-88df-de63ceea3402.gif)

***Gif of menu tabbing closing the menu***
![devtools-menu-console-esc-bug-fixed-demo-tabbing](https://user-images.githubusercontent.com/309310/72119079-0a131880-3308-11ea-81ee-f6665e125dbe.gif)

**Related Issues/PRs**
#9433

[1]: https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190207/examples/menu-button/menu-button-links.html
[2]: https://github.com/jsdom/jsdom/issues/2708
[3]: https://github.com/microsoft/TypeScript/issues/30716

johnemau commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -744,6 +747,21 @@ class DropDown {
+  onMenuFocusOut(e) {
+    // The focusout event is not supported in our current version of typescript (3.5.3)
+    // https://github.com/microsoft/TypeScript/issues/30716
+    const focusEvent = /** @type {FocusEvent} */ (e);
```
Is there appetite to upgrade typescript?

connorjclark replied:
I'm always up for upgrading TS, but we may not be at critical mass yet ... @patrickhulce what do you think?

johnemau commented:
I am going to look into adding a unit test for onMenuFocusOut to improve code coverage per codecov results.

connorjclark commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -744,6 +747,21 @@ class DropDown {
+    // The focusout event is not supported in our current version of typescript (3.5.3)
+    // https://github.com/microsoft/TypeScript/issues/30716
+    const focusEvent = /** @type {FocusEvent} */ (e);
+    const focusedEl = /** @type {?HTMLElement} */ (focusEvent.relatedTarget);
```
is this nullable?

johnemau replied:
Yes, relatedTarget might be null.
https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/FocusEvent

connorjclark replied:
ah, I see, this is another thing that newer TS has corrected.

connorjclark commented:
> I have just finished writing a few unit tests and will add this to this PR but I am happy to follow up with UX tests if you think they would be valuable.

If you feel that covers the new behavior well enough that is fine. Was just suggesting an alternative in case unit tests weren't possible.

> I do not know about the Print preview issue, do you have a link to the issue or repro steps? I will take a look.

no crbug, just something I noticed. Printing doesn't seem to be a feature anyone uses so no big deal, was just curious. If you choose the Print option from the tool menu and don't see this tool menu in the Print results, then you've fixed it :)

johnemau commented:
> If you feel that covers the new behavior well enough that is fine. Was just suggesting an alternative in case unit tests weren't possible.

I gotcha, the unit tests I added are good enough coverage.

> no crbug, just something I noticed. Printing doesn't seem to be a feature anyone uses so no big deal, was just curious. If you choose the Print option from the tool menu and don't see this tool menu in the Print results, then you've fixed it :)

Unfortunately this change did not fix Print preview :(

![image](https://user-images.githubusercontent.com/309310/72196969-131bec80-33d1-11ea-9cf5-78274b33dd3a.png)

But this looks like a CSS issue, the menu is hidden by `visibility:hidden` and so I fixed it locally by adding:
 ```css
@media print { 
  .lh-tools__dropdown { 
      /* the drop down is not supported on paper ;) */
      display: none; 
  }
```

I will get a PR out with that fix shortly.

johnemau commented:
> I will get a PR out with that fix shortly.

Done: #10216

connorjclark commented on lighthouse-core/test/report/html/renderer/report-ui-features-test.js:
```diff
@@ -395,5 +395,50 @@ describe('ReportUIFeatures', () => {
+
+    describe('onMenuFocusOut', () => {
+      it('should toggle active class when focus relatedTarget is null', () => {
+        dropDown._toggleEl.click();
```
nit: move these first two lines to `beforeEach` scoped to this new `describe`?

johnemau replied:
Good idea, thanks! Done.

johnemau replied:
Good idea, thanks! Done.

connorjclark commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -744,6 +747,21 @@ class DropDown {
+   * @param {Event} e
+   */
+  onMenuFocusOut(e) {
+    // The focusout event is not supported in our current version of typescript (3.5.3)
```
```suggestion
    // TODO: The focusout event is not supported in our current version of typescript (3.5.3)
```

Let's just handle this later.

johnemau replied:
Done!
